Added dedicated support for hyperdb

JimmyBoh opened the pull request:
As per the discussion on #11 here is dedicated support for hyperdb. There are 3 major requirements, the first is passing the correct `id` to hyperdiscovery, the second is overriding the stream to include the local id in the `userData` (exposed by hypercore-protocol), and the third is authenticating new connections based on that `remoteUserData`. This achieves all three only when it is a hyperdb instance (or when it looks like one, any recommendations for testing if hyperdb would be appreciated) in the least intrusive way.

I left some `debug` log calls in there commented out, just to indicate what I've done locally, but I can remove those if this PR is something that could be merged.

The tests do use a timeout to delay because I couldn't find a good event to bind to for indicating when two databases are in sync. Arbitrary delays are not ideal but it seems to work fine.

@pvh wanted to see this patch, so if it is too divergent then no worries!

JimmyBoh commented:
Yep, timeout bites back! I'll try re-working the tests later.

karissa commented:
Awesome! If there isn't a way to see when two databases in sync, I think there ought to be! @mafintosh ?

JimmyBoh commented:
So after watching all of the events that I can see from hyperdb, hyperdiscovery, and discovery-swarm it really just seems like `download` is key, but it does fire multiple times. So I added a simple debounce helper function and just wait until the last `download` call. That and I changed the test to update the databases sequentially instead of at the same time. As long as hyperdiscovery can get two hyperdb instances connected and reading/writing then I think the feature is a success.

JimmyBoh commented:
Hey @karissa, any changes you see that I should make? 

@mafintosh Is there a better way to check for a "sync completion" event? Thanks!

mafintosh commented on index.js:
```diff
@@ -13,25 +14,45 @@ function HyperdriveSwarm (archive, opts) {
+  // Authorize peer connections if download is enabled...
+  this.downloading && this.swarm.on('connection', function (peer) {
+    if (peer.remoteUserData) {
+      archive.authorize(peer.remoteUserData, function () {
```
this doesn't seem right? are you auto authenticating writers on every connect?

JimmyBoh replied:
Hmm, good point. What is the best approach for checking if a user is already authorized?

karissa replied:
I think you can just remove this line

JimmyBoh replied:
In order for outsiders to write I think it is need, but I may just have over-complicated it. I might have time this week, I'll take another look at it.

karissa replied:
Yes, but in some cases, the original creator might want the db to be read-only, or in case the key gets leaked, prevents spam from unwanted writers. So, writers are added explicitly.

karissa replied:
Someone can still extend this module by adding this code in the `.on('connection'` event.

substack replied:
this .authorize() trick with userData is a fantastic hack that unlocks new potential

JimmyBoh replied:
@karissa Could I put it behind an option flag? So you have to opt-in to the auto-authorize? @mafintosh Is there any way to avoid authorizing the same people multiple times.

karissa replied:
@JimmyBoh I think auto-authorization is out of the scope of this module

JimmyBoh replied:
Sorry for the silence, I am working on removing the auto-authorization aspect now.

Frando commented on index.js:
```diff
@@ -13,25 +14,45 @@ function HyperdriveSwarm (archive, opts) {
   this.live = !(opts.live === false)
+  this.authorize = !(opts.authorize === false)
+
+  var isHyperdbInstance = !!(archive.get && archive.put && archive.replicate && archive.authorize)
```
Something like this will be needed in quite a few places, we also stumbled upon this need when making hyper-graph-db compatible with subhyperdb. What about adding a property `isHyperdb` to the hyperdb object to make this simpler and more error-proof?

karissa replied:
+1 @mafintosh

karissa replied:
@Frando do you have any time to open a PR on hyperdb for this?

micahscopes replied:
out of curiosity, why not use `this instanceOf HyperDB`?

karissa replied:
@micahscopes ah that wouldn't work because this module doesn't `require` hyperdb or hypercore

JimmyBoh replied:
@karissa is spot on, good old duck typing. 🦆

JimmyBoh commented:
There, those recent changes remove the auto-authorize logic and simplify the test. I can add more "guts" to the test if needed.

karissa commented:
looks great! will publish this as the next major version.

karissa commented:
8.0.0

JimmyBoh commented:
Great, thanks @karissa!
